Mathlib/RingTheory/Localization/Defs.lean

```diff
@@ -931,3 +933,71 @@ theorem isDomain_localization {M : Submonoid A} (hM : M ≤ nonZeroDivisors A) :
 end IsLocalization
 
 end CommRing
+
+section
+
+open IsLocalization
+
+variable (S : Type*) [CommRing S] {ι : Type*} (R' S' : ι → Type*) [∀ i, CommRing (R' i)]
+  [∀ i, CommRing (S' i)] [Algebra (∀ i, R' i) S] [∀ i, Algebra (R' i) (S' i)]
+  {M : Submonoid (∀ i, R' i)} [sloc : IsLocalization M S]
+  {M' : ∀ i, Submonoid (R' i)} [isloc : ∀ i, IsLocalization (M' i) (S' i)]
+
+/-- Let `M` be a submonoid of a direct product of commutative rings `R' i`, and let `M' i` denote
+the projection of `M` onto each corresponding factor. Given a ring homomorphism from the direct
+product `∏ R' i` to the product of the localizations of each `R' i` at `M' i`, every `y : M`
+maps to a unit under this homomorphism.
+-/
+lemma isUnit_of_product_of_localizations (h : ∀ i, (M' i) = M.map (Pi.evalRingHom R' i)) (y : M) :
+    IsUnit ((Pi.ringHom fun i ↦ (algebraMap (R' i) (S' i)).comp (Pi.evalRingHom R' i)) y) := by
+  set f' : ∀i, (∀i, R' i) →+* S' i :=
+    fun i ↦ RingHom.comp (algebraMap (R' i) (S' i)) (Pi.evalRingHom R' i)
+  refine isUnit_iff_exists_inv.mpr ?_
+  use fun i ↦ Ring.inverse ((f' i) y)
+  rw [mul_comm]
+  exact funext (fun i ↦ Ring.inverse_mul_cancel ((f' i) y) ((isloc i).map_units (S' i) ⟨y.1 i,
+    by rw [h i]; exact Submonoid.mem_map_of_mem (Pi.evalRingHom R' i) y.2⟩))
```

Comment:
This line is pretty complicated. Could we break out the `by ... ` block into a preliminary `have`?

Reply:
```suggestion
  exact funext fun i ↦ Ring.inverse_mul_cancel (f' i y) ((isloc i).map_units (S' i) ⟨y.1 i,
    by rw [h i]; exact Submonoid.mem_map_of_mem (Pi.evalRingHom R' i) y.2⟩)
```
or maybe
```suggestion
  exact funext fun i ↦ Ring.inverse_mul_cancel (f' i y) ((isloc i).map_units (S' i) ⟨y.1 i, ?_⟩)
  rw [h i]; exact Submonoid.mem_map_of_mem (Pi.evalRingHom R' i) y.2
```